proxy/stats_get_prediction.go

```diff
@@ -0,0 +1,345 @@
+	// TODO: Improve this.
+	// Counts up for env->login->stats/get, to kick off Async StatsGet predictive cache
+	statsGetPredictionCounter int
+	statsGetTasks             map[string]*StatsGetTask
+}
+
+func (s *StatsGetPrediction) HandleCatchallPath(path string) {
+	if !s.enable {
+		return
+	}
+	if s.statsGetPredictionCounter > 0 && path != "/statistics/get" {
+		fmt.Println("Done looking up stats")
+		s.statsGetPredictionCounter = 0
+	}
+}
+
+func (s *StatsGetPrediction) HandleGetEnv() {
+	if !s.enable {
+		return
+	}
+	s.statsGetPredictionCounter = 1
+}
+
+func (s *StatsGetPrediction) HandleLoginData(login []byte) {
+	if !s.enable {
+		return
+	}
+	if s.statsGetPredictionCounter == 1 {
+		s.statsGetPredictionCounter = 2
+		s.ParseLoginPrefix(login)
+	}
+}
+
+// Proxy getstats
+func (s *StatsGetPrediction) HandleGetStats(w http.ResponseWriter, r *http.Request) bool {
+	if !s.enable {
+		return false
+	}
+	if len(s.loginPrefix) > 0 && s.statsGetPredictionCounter == 2 {
+		s.AsyncGetStats()
+		s.statsGetPredictionCounter = 3
+	}
+	if len(s.loginPrefix) > 0 && s.statsGetPredictionCounter == 3 {
+		bodyBytes, _ := ioutil.ReadAll(r.Body)
+		r.Body.Close() //  must close
+		r.Body = ioutil.NopCloser(bytes.NewBuffer(bodyBytes))
+		req := string(bodyBytes)
+		if task, ok := s.statsGetTasks[req]; ok {
+			resp := <-task.response
+			if resp == nil {
+				fmt.Println("Cache Error!")
+				delete(s.statsGetTasks, req)
+				return false
+			}
+			defer resp.Body.Close()
+			// Copy headers
+			for name, values := range resp.Header {
+				w.Header()[name] = values
+			}
+			w.WriteHeader(resp.StatusCode)
+			reader := io.TeeReader(resp.Body, w) // For dumping API payloads
+			_, err := io.ReadAll(reader)
+			if err != nil {
+				fmt.Println(err)
+			}
+			delete(s.statsGetTasks, req)
+			return true
+		}
+		fmt.Println("Cache miss! " + req)
+		return false
+
+	}
+	return false
+}
+
+func (s *StatsGetPrediction) ParseLoginPrefix(loginRet []byte) {
+	s.loginPrefix = hex.EncodeToString(loginRet[60:79]) + hex.EncodeToString(loginRet[2:16])
+}
+
+func (s *StatsGetPrediction) BuildStatsReqBody(login string, req string) string {
+	/*
+
+		Get Stats Call Analysis
+		E.g.
+		data=9295xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx02a5302e302e350396a007ffffffff^@
+
+		1.
+		"data="
+		length=5
+
+		2. Header?
+		9295
+		length=2
+
+
+		3. Login 1?
+		index in login response=60
+		length=19
+
+		4. Login 2?
+		index in login response=2
+		length=14
+
+		5. Divider?
+		02a5
+		l=2
+
+		6. Version?
+		302e302e35 (0.0.5)
+		l=5
+
+		7. Divider2?
+		0396
+		l=2
+
+		8. Specific call
+		e.g. a007ffffffff , Confirm that this stays between users
+		l=6
+
+
+		9=End
+		\0
+		l=1
+	*/
+
+	body := "data=" +
+		"9295" + // Header
+		login +
+		"02a5" + // Divider
+		"302e302e35" + // 0.0.5
+		"0396" + // Divider 2
+		req +
+		"\x00" // End
+
+	return body
+}
+
+func (s *StatsGetPrediction) ProcessStatsQueue(queue chan *StatsGetTask) {
+	client := http.Client{
```

Comment:
Clients are goroutine safe. You can safely re-use the one from the proxy to make sure all the requests fit within the single connection.

Reply:
Making 5x clients makes 5x independent connections. ASW servers may not like users having 5x more connections than vanilla.

Reply:
This code does rely on making simultaneous requests.
I can use a single client (see update) but still it relies on a max connection pool of at least size 4 in order to get optimal time. With a connection pool of 4 and 5 threads I was able to get 7s
8s for pool of 3 with 4 threads
9s for pool of 2 with 3 threads

>poolSize+1 threads deadlocked

Reply:
Oops, forgot that was the whole point of this.